### PR TITLE
Revert #762 (Grade upgrade) and add CI checking of publishing flows

### PR DESCRIPTION
**Pull request by lazaroclapp**
A recent upgrade to our gradle version (#762) silently broke our publishing flows.

In particular, `publish` and `publishToMavenLocal` are broken on Java 8, resulting in the following error:

```
Execution failed for task ':annotations:publishMavenPublicationToMavenLocal'.
> Failed to publish publication 'maven' to repository 'mavenLocal'
   > Invalid publication 'maven': artifact file does not exist: '/home/runner/work/NullAway/NullAway/annotations/build/publications/maven/module.json.asc'
```

See this [CI job run](https://github.com/uber/NullAway/actions/runs/5261437721/jobs/9509542067?pr=767).
There are further errors when the release is signed.

Below, we revert the gradle upgrade, and also add a CI job running `publishToMavenLocal` to prevent
future regressions (We can't trivially test signing these publications in a secure manner, unfortunately, 
so we skip that for now. If that specifically breaks in the future, we could provide a test gpg key for signing
in CI).

### CODE REVIEW

File: .github/workflows/continuous-integration.yml [hunk 1]
```diff
@@ -77,6 +77,14 @@ jobs:
           arguments: coveralls
         continue-on-error: true
         if: runner.os == 'Linux' && matrix.java == '11' && matrix.epVersion == '2.19.1' && github.repository == 'uber/NullAway'
+      - name: Test publishToMavenLocal flow
+        env:
+          ORG_GRADLE_PROJECT_epApiVersion: ${{ matrix.epVersion }}
+          ORG_GRADLE_PROJECT_VERSION_NAME: '0.0.0.1-LOCAL'
+        uses: gradle/gradle-build-action@v2
+        with:
+          arguments: publishToMavenLocal -x signMavenPublication -x signShadowPublication
```
Comment (lazaroclapp) [inline .github/workflows/continuous-integration.yml:86]: 1. I go back and forth about whether to include `clean` here. On the one hand, that's what we do locally. On the other, it's a waste of CI resources to build twice and we don't really mind if this is more brittle than locally, unless it's super noisy.
2. Not sure if it's worth setting up dummy signing credentials. Right now, the lack of them means we aren't testing `signMavenPublication` or `signShadowPublication`, and locally, before the revert, I did see errors related to those when running `publish` (but a different error shows up anyways with the command above, so we would have caught this PR either way)

Comment (msridhar) [reply .github/workflows/continuous-integration.yml:86]: 1. I think it's fine not to run `clean`.
2. Don't we have credentials set up somewhere to push snapshot builds?  Or are those unsigned?

**Review by msridhar — APPROVED** [review]
Can we open an issue to track the broken publications on Gradle 8.1.1?  We need to fix that eventually.  Otherwise this looks good to land, comment is non-blocking

File: .github/workflows/continuous-integration.yml (hunk 1 above)
Comment (lazaroclapp) [reply .github/workflows/continuous-integration.yml:86]: Snapshot builds are unsigned, which is also why we didn't see any issue with the upgrade, generally snapshot builds are different. This check also didn't trigger the error until I added `ORG_GRADLE_PROJECT_VERSION_NAME: '0.0.0.1-LOCAL'` making a (meaninglessly numbered) non-SNAPSHOT version.

**Pull request merged**